pulsar-client/src/main/java/org/apache/pulsar/client/impl/MultiTopicsConsumerImpl.java

```diff
@@ -1059,29 +1059,28 @@ private void doSubscribeTopicPartitions(Schema<T> schema,
 
             CompletableFuture<Consumer<T>> subFuture = new CompletableFuture<>();
 
-            consumers.compute(topicName, (key, existingValue) -> {
-                if (existingValue != null) {
-                    String errorMessage = String.format("[%s] Failed to subscribe for topic [%s] in topics consumer. "
-                            + "Topic is already being subscribed for in other thread.", topic, topicName);
-                    log.warn(errorMessage);
-                    subscribeResult.completeExceptionally(new PulsarClientException(errorMessage));
-                    return existingValue;
-                } else {
-                    internalConfig.setStartPaused(paused);
-                    ConsumerImpl<T> newConsumer = createInternalConsumer(internalConfig, topicName,
-                            -1, subFuture, createIfDoesNotExist, schema);
-
-                    synchronized (pauseMutex) {
+            synchronized (pauseMutex) {
```

Comment:
An alternative solution could be something like

```java
            internalConfig.setStartPaused(paused);
            ConsumerImpl<T> newConsumer = createInternalConsumer(internalConfig, topicName,
                    -1, subFuture, createIfDoesNotExist, schema);
            synchronized (pauseMutex) {
                if (paused) {
                    newConsumer.pause();
                } else {
                    newConsumer.resume();
                }
                ConsumerImpl<T> existingValue = consumers.putIfAbsent(topicName, newConsumer);
                if (existingValue != null) {
                    String errorMessage = String.format("[%s] Failed to subscribe for topic [%s] in topics consumer. "
                            + "Topic is already being subscribed for in other thread.", topic, topicName);
                    log.warn(errorMessage);
                    subscribeResult.completeExceptionally(new PulsarClientException(errorMessage));
                }
            }
```
This block is similar to others in the class. However, the primary issue is that we'd need to close the consumer if it is already created. Speaking of closing consumers, I actually think we might have a leak in this code:

https://github.com/apache/pulsar/blob/e956db729f5098ff319237fd4220dbfb234c1b18/pulsar-client/src/main/java/org/apache/pulsar/client/impl/MultiTopicsConsumerImpl.java#L1044-L1053

Reply:
@michaeljmarshall  wouldn't there also be a need to cleanup `newConsumer` if it doesn't get used at all?

Reply:
Oh.. should we clean the code like this ?

https://github.com/heesung-sn/pulsar/commit/0c415d7a6df9c12f1e4a78e640b2446230b96136

Reply:
@lhotari - see the second half of my comment.

Reply:
> Oh.. should we clean the code like this ?
> 
> [heesung-sn@0c415d7](https://github.com/heesung-sn/pulsar/commit/0c415d7a6df9c12f1e4a78e640b2446230b96136)

That approach would work. Do you know how expensive the consumer creation is? We could also use the class's internal executor to do it in a way that doesn't block the calling threads.

Reply:
IMO, though it creates many other objects, the most expensive op from consumer creation is `grabCnx();`,  but I believe this connection creation is an async operation.